File: samples/perl/GoogleTests.pl
package Google::Test;
use strict;
use warnings;
use Selenium::Remote::Driver;
use base qw(Test::Class);
use Test::More;

my $SERVER_URL_LOCAL = 'http://127.0.0.1:4723/wd/hub';
my $USER_NAME = 'tester01';
my $API_KEY = '13e36639-92e3-411b-a067-3457b5dea573';
my $HOST_NAME = '192.168.74.51';
my $PORT = '3001';
my $GOOGLE_URL = 'https://mail.google.com';
my $GOOGLE_URL_HOMEPAGE = 'https://mail.google.com/mail/mu/mp/4/#tl/priority/%5Esmartlabel_personal';
my $TRASH_URL = 'https://mail.google.com/mail/mu/mp/4/#tl/Trash';
my $SENT_EMAIL_URL='https://mail.google.com/mail/mu/mp/68/#tl/Sent%20Mail';
my $PRIMARY_EMAIL_URL='https://mail.google.com/mail/mu/mp/68/#tl/priority/%5Esmartlabel_personal';
my $EMAIL_ADDRESS1 = 'krypton.portal@gmail.com';
my $EMAIL_ADDRESS2 = 'krypton.portal2@gmail.com';
my $PASSWORD = 'Admin@123456';
my $SUBJECT1 = 'It is a subject';
my $BODY1 = 'It is a body';
my $MSG_NOEMAIL = 'You have no mail here.';
my $MSG_NOEMAIL_PRIMARY1 = 'You have no mail.';
my $MSG_NOEMAIL_PRIMARY2 = 'Please enjoy your day!';
my $driver;
my $caps = {
    browserName       => "Chrome",
    deviceName        => "Nexus 5",
    platformName      => "ANDROID",
    platformVersion   => "5.1.1"
};

sub before : Test(setup){
  my $SERVER_URL = "http://" . $USER_NAME . ":" . $API_KEY . "\@" . $HOST_NAME . ":" . $PORT . "/v1/tests";
  print $SERVER_URL ."\n";
  $driver = Selenium::Remote::Driver->new_from_caps(
     remote_server_addr   => $HOST_NAME,
     port                 => $PORT,
     url                  => $SERVER_URL,
     desired_capabilities => $caps
  );

  $driver->set_implicit_wait_timeout(5000);
  $driver->set_timeout("page load",10000);
  $driver->delete_all_cookies();
}

sub should_not_accept_empty_or_invalid_email() : Test(2){
  print "should not accept empty or invalid email \n";
  $driver->get($GOOGLE_URL);
  $driver->find_element('next','id')->click();
  sleep 3;
  my $getError = $driver->find_element('errormsg_0_Email','id')->get_text();
  ok($getError, "Please enter your email.");
  $driver->find_element('Email','id')->send_keys('invalid_email@where.about');
  $driver->find_element('next','id')->click();
  sleep 3;
  ok($getError, "Sorry, Google doesn't recognize that email. Create an account using that address?");
}

sub should_login_and_compose_email_successfully() : Test(18) {
  print "should accept valid credentials \n";
  &login($EMAIL_ADDRESS1, $PASSWORD);
  my $getCurrentURL= $driver->get_current_url();
  my $expectedURL="https://mail.google.com/mail";
  like($getCurrentURL, qr/$expectedURL/,"https://mail.google.com/mail");
  print "should accept compose email successfully \n";
  &compose_email($EMAIL_ADDRESS2, $SUBJECT1, $BODY1);
  &verify_sent_email($SUBJECT1, $BODY1, $EMAIL_ADDRESS1, $EMAIL_ADDRESS2);
  &delete_sent_email();
  &delete_trash_email();
  &logout();
  &login($EMAIL_ADDRESS2, $PASSWORD);
  &verify_primary_email($SUBJECT1, $BODY1, $EMAIL_ADDRESS1, $EMAIL_ADDRESS2);
  &delete_primary_email();
  &delete_trash_email();
}

sub should_accept_valid_credentials(){
  print "should accept valid credentials \n";
  &login($EMAIL_ADDRESS1, $PASSWORD);
  my $getCurrentURL= $driver->get_current_url();
  my $expectedURL="https://mail.google.com/mail";
  like($getCurrentURL, qr/$expectedURL/,"https://mail.google.com/mail");
}

sub login(){
  print "call sub - login \n";
  my($username,$password)= @_;
  $driver->get($GOOGLE_URL);
  $driver->find_element('Email','id')->send_keys($username);
  $driver->find_element('next','id')->click();
  sleep 1;
  $driver->find_element('Passwd','id')->send_keys($password);
  $driver->find_element('signIn','id')->click();
  sleep 10;
}

sub logout(){
  print "call sub - logout \n";
  $driver->get($PRIMARY_EMAIL_URL);
  sleep 2;
  $driver->find_element('//div[@aria-label=\'Menu\']','xpath')->click();
  sleep 2;
  $driver->find_element('//div[contains(@class,\'V Y Rx Kg\')]','xpath')->click();
  sleep 2;
  $driver->find_element('//button[@id=\'signout\']','xpath')->click();
  sleep 5;
  $driver->get($GOOGLE_URL);
  $driver->delete_all_cookies();
  $driver->refresh();
  }

sub compose_email(){
  print "call sub - compose email \n";
  my($toEmail, $subject, $body)=@_;
  $driver->find_element('//div[@id=\'views\']//div[@aria-label=\'Compose\']','xpath')->click();
  sleep 1;
  $driver->find_element('//div[@id=\'cmcc_composeto\']//input[@id=\'composeto\']','xpath')->send_keys($toEmail);
  $driver->find_element('//input[@id=\'cmcsubj\']','xpath')->send_keys($subject);
  $driver->find_element('cmcbody','id')->send_keys($body);
  sleep 2;
  $driver->find_element('//div[@id=\'views\']//div[text()=\'Send\']','xpath')->click();
  sleep 1;
}

sub verify_sent_email(){
  print "call sub - verify sent email \n";
  my ($subject, $body, $sendEmail, $toEmail)=@_;
  $driver->find_element('//div[@aria-label=\'Menu\']','xpath')->click();
  sleep 2;
  $driver->find_element('//span[text()=\'Sent Mail\']','xpath')->click();
  sleep 2;
  my $getText = $driver->find_element('//div[@class=\'fm\']','xpath')->get_text();
  like($getText, qr/$subject/,$subject);
  like($getText, qr/$body/,$body);

  $driver->find_element('//div[contains(@class,\'m\')]//div[@role=\'listitem\'][1]','xpath')->click();
  sleep 1;
  $driver->find_element('//div[contains(@class,\'V j hj\') and text()=\'Details\']','xpath')->click();
  sleep 1;
  my $getSubject = $driver->find_element('//span[@class=\'kj\']/span','xpath')->get_text();
  my $getBody = $driver->find_element('//div[@class=\'Hi\']','xpath')->get_text();
  my $getSendEmail = $driver->find_element('//div[@class=\'Kg\']/span','xpath')->get_text();
  my $getToEmail = $driver->find_element('//div[@class=\'Kh\']//span[@class=\'Hh\']//span[@class=\'Kg ph\']','xpath')->get_text();

  like($getSubject, qr/$subject/,$subject);
  like($getBody, qr/$body/,$body);
  like($getSendEmail, qr/$sendEmail/,$sendEmail);
  like($getToEmail, qr/$toEmail/,$toEmail);
}

sub verify_primary_email(){
  print "call sub - verify primary email \n";
  my ($subject, $body, $sendEmail, $toEmail)=@_;
  $driver->get($GOOGLE_URL_HOMEPAGE);
  sleep 2;
  my $getText = $driver->find_element('//div[@class=\'fm\']','xpath')->get_text();

  like($getText, qr/$subject/,$subject);
  like($getText, qr/$body/,$body);

  $driver->find_element('//div[contains(@class,\'fm\')]//div[@role=\'listitem\'][1]','xpath')->click();
  sleep 1;
  $driver->find_element('//div[contains(@class,\'V j hj\') and text()=\'Details\']','xpath')->click();
  sleep 1;
  my $getSubject = $driver->find_element('//span[@class=\'kj\']/span','xpath')->get_text();
  my $getBody = $driver->find_element('//div[@class=\'Hi\']','xpath')->get_text();
  my $getSendEmail = $driver->find_element('//div[@class=\'Kg\']/span','xpath')->get_text();
  my $getToEmail = $driver->find_element('//div[@class=\'Kh\']//span[@class=\'Hh\']//span[@class=\'Kg ph\']','xpath')->get_text();

  like($getSubject, qr/$subject/,$subject);
  like($getBody, qr/$body/,$body);
  like($getSendEmail, qr/$sendEmail/,$sendEmail);
  like($getToEmail, qr/$toEmail/,$toEmail);
}

sub delete_sent_email(){
  print "call sub - delete sent email \n";
  $driver->get($SENT_EMAIL_URL);
  sleep 2;
  $driver->find_element('//div[@class=\'V j cb Ol\'][1]','xpath')->click();
  sleep 1;
  $driver->find_element('//div[@class=\'M j T b hc  Vm Im\']//div[@class=\'V j Xd\']','xpath')->click();
  sleep 1;
  my $noEmailSentEmail = $driver->find_element('//div[@class=\'Wl\']','xpath')->get_text();
  ok($MSG_NOEMAIL,$noEmailSentEmail);
}

sub delete_trash_email(){
  print "call sub - delete trash email \n";
  $driver->get($TRASH_URL);
  sleep 2;
  $driver->find_element('//div[@class=\'V j cb Ol\'][1]','xpath')->click();
  sleep 1;
  $driver->find_element('//div[@class=\'M j T b hc q m\']//div[text()=\'Delete forever\']','xpath')->click();
  sleep 1;
  my $noEmailTrash = $driver->find_element('//div[@class=\'Wl\']','xpath')->get_text();
  ok($MSG_NOEMAIL,$noEmailTrash);
}

sub delete_primary_email(){
  print "call sub - delete primary email \n";
  $driver->get($PRIMARY_EMAIL_URL);
  sleep 2;
  $driver->find_element('//div[@class=\'V j cb Ol\'][1]','xpath')->click();
  sleep 1;
  $driver->find_element('//div[@class=\'M j T b hc  Vm Im\']//div[@class=\'V j Xd\']','xpath')->click();
  sleep 1;
  my $noEmailPrimary = $driver->find_element('//div[@class=\'Wl\']','xpath')->get_text();

  like($noEmailPrimary, qr/$MSG_NOEMAIL_PRIMARY1/,$MSG_NOEMAIL_PRIMARY1);
  like($noEmailPrimary, qr/$MSG_NOEMAIL_PRIMARY2/,$MSG_NOEMAIL_PRIMARY2);
}

sub after : Test(teardown) {
  $driver->quit;
  sleep 10;
 }

done_testing();

Test::Class->runtests;

File: samples/php/GoogleTests.php
<?php
class SampleTest extends PHPUnit_Framework_TestCase {

  protected $driver;
  protected $SERVER_URL_LOCAL = 'http://127.0.0.1:4723/wd/hub';
  protected $USER_NAME = 'tester01';
  protected $API_KEY = '13e36639-92e3-411b-a067-3457b5dea573';
  protected $HOST_NAME = 'test.kobiton.com';
  protected $PORT = '3001';
  protected $GOOGLE_URL = 'https://mail.google.com';
  protected $GOOGLE_URL_HOMEPAGE = 'https://mail.google.com/mail/mu/mp/4/#tl/priority/%5Esmartlabel_personal';
  protected $TRASH_URL = 'https://mail.google.com/mail/mu/mp/4/#tl/Trash';
  protected $SENT_EMAIL_URL='https://mail.google.com/mail/mu/mp/68/#tl/Sent%20Mail';
  protected $PRIMARY_EMAIL_URL='https://mail.google.com/mail/mu/mp/68/#tl/priority/%5Esmartlabel_personal';
  protected $EMAIL_ADDRESS1 = 'krypton.portal@gmail.com';
  protected $EMAIL_ADDRESS2 = 'krypton.portal2@gmail.com';
  protected $PASSWORD = 'Admin@123456';
  protected $SUBJECT1 = 'It is a subject';
  protected $BODY1 = 'It is a body';
  protected $MSG_NOEMAIL = 'You have no mail here.';
  protected $MSG_NOEMAIL_PRIMARY1 = 'You have no mail.';
  protected $MSG_NOEMAIL_PRIMARY2 = 'Please enjoy your day!';

  public function setUp() {
    $capabilities=array(
    'browserName' => 'Chrome',
    'platformVersion' => '5.1.1',
    'platformName' => 'Android',
    'deviceName' => 'Nexus 5'
    );
    print_r($capabilities);
    $SERVER_URL = $this->USER_NAME . ':' . $this->API_KEY . '@' . $this->HOST_NAME .':' . $this->PORT . '/wd/hub';
    print_r($SERVER_URL);
    $this->driver = RemoteWebDriver::create($SERVER_URL, $capabilities);
    $this->driver->manage()->timeouts()->implicitlyWait(3000);
    $this->driver->manage()->deleteAllCookies();
  }

  public function test_should_not_accept_empty_or_invalid_email() {
    print_r("should not accept empty or invalid email");
    $this->driver->get($this->GOOGLE_URL);
    $this->driver->findElement(WebDriverBy::id("next"))->click();
    sleep(1);
    $getText = $this->driver->findElement(WebDriverBy::id("errormsg_0_Email"))->getText();
    $this->assertEquals($getText, 'Please enter your email.');
    $this->driver->findElement(WebDriverBy::id("Email"))->sendKeys('invalid_email@where.about');
    $this->driver->findElement(WebDriverBy::id("next"))->click();
    sleep(1);
    $getText = $this->driver->findElement(WebDriverBy::id("errormsg_0_Email"))->getText();
    $this->assertEquals($getText, "Sorry, Google doesn't recognize that email. Create an account using that address?");
  }

  public function test_should_accept_valid_credentials() {
    print_r("should accept valid credentials");
    $this->login($this->EMAIL_ADDRESS1, $this->PASSWORD);
    $this->assertContains('https://mail.google.com/mail',$this->driver->getCurrentURL());
  }

  public function test_should_compose_email_successfully()
  {
    print_r("should accept compose email successfully");
      $this->login($this->EMAIL_ADDRESS1, $this->PASSWORD);
      $this->compose_email($this->EMAIL_ADDRESS2, $this->SUBJECT1, $this->BODY1);
      $this->verify_sent_email($this->SUBJECT1, $this->BODY1, $this->EMAIL_ADDRESS1, $this->EMAIL_ADDRESS2);
      $this->delete_sent_email();
      $this->delete_trash_email();
      $this->logout();
      $this->login($this->EMAIL_ADDRESS2, $this->PASSWORD);
      $this->verify_primary_email($this->SUBJECT1, $this->BODY1, $this->EMAIL_ADDRESS1, $this->EMAIL_ADDRESS2);
      $this->delete_primary_email();
      $this->delete_trash_email();
  }

  protected function login($username, $password)
  {
    $this->driver->get($this->GOOGLE_URL);
    $this->driver->findElement(WebDriverBy::id("Email"))->sendKeys($username);
    $this->driver->findElement(WebDriverBy::id("next"))->click();
    sleep(1);
    $this->driver->findElement(WebDriverBy::id("Passwd"))->sendKeys($password);
    $this->driver->findElement(WebDriverBy::id("signIn"))->click();
    sleep(10);
  }

  protected function logout()
  {
    $this->driver->get($this->PRIMARY_EMAIL_URL);
    $this->driver->findElement(WebDriverBy::xpath("//div[@aria-label='Menu']"))->click();
    sleep(1);
    $this->driver->findElement(WebDriverBy::xpath("//div[contains(@class,'V Y Rx Kg')]"))->click();
    sleep(1);
    $this->driver->findElement(WebDriverBy::xpath("//button[@id='signout']"))->click();
    sleep(5);
    $this->driver->get($this->GOOGLE_URL);
    $this->driver->manage()->deleteAllCookies();
    $this->driver->navigate()->refresh();
  }

  protected function compose_email($toEmail, $subject, $body)
  {
    $this->driver->findElement(WebDriverBy::xpath("//div[@id='views']//div[@aria-label='Compose']"))->click();
    sleep(1);
    $this->driver->findElement(WebDriverBy::xpath("//div[@id='cmcc_composeto']//input[@id='composeto']"))->sendKeys($toEmail);
    $this->driver->findElement(WebDriverBy::xpath("//input[@id='cmcsubj']"))->sendKeys($subject);
    $this->driver->findElement(WebDriverBy::id("cmcbody"))->sendKeys($body);
    sleep(2);
    $this->driver->findElement(WebDriverBy::xpath("//div[@id='views']//div[text()='Send']"))->click();
    sleep(1);
  }

  protected function verify_sent_email($subject, $body, $sendEmail, $toEmail)
  {
    $this->driver->findElement(WebDriverBy::xpath("//div[@aria-label='Menu']"))->click();
    sleep(2);
    $this->driver->findElement(WebDriverBy::xpath("//span[text()='Sent Mail']"))->click();
    sleep(2);
    $getText = $this->driver->findElement(WebDriverBy::xpath("//div[@class='fm']"))->getText();

    $this->assertContains($subject,$getText);
    $this->assertContains($body,$getText);

    $this->driver->findElement(WebDriverBy::xpath("//div[contains(@class,'m')]//div[@role='listitem'][1]"))->click();
    sleep(1);
    $this->driver->findElement(WebDriverBy::xpath("//div[contains(@class,'V j hj') and text()='Details']"))->click();
    $getSubject = $this->driver->findElement(WebDriverBy::xpath("//span[@class='kj']/span"))->getText();
    $getBody = $this->driver->findElement(WebDriverBy::xpath("//div[@class='Hi']"))->getText();
    $getSendEmail = $this->driver->findElement(WebDriverBy::xpath("//div[@class='Kg']/span"))->getText();
    $getToEmail = $this->driver->findElement(WebDriverBy::xpath("//div[@class='Kh']//span[@class='Hh']//span[@class='Kg ph']"))->getText();

    $this->assertContains($subject,$getSubject);
    $this->assertContains($body,$getBody);
    $this->assertContains($sendEmail,$getSendEmail);
    $this->assertContains($toEmail,$getToEmail);
  }

  protected function verify_primary_email($subject, $body, $sendEmail, $toEmail)
  {
    $this->driver->get($this->GOOGLE_URL_HOMEPAGE);
    sleep(2);
    $getText = $this->driver->findElement(WebDriverBy::xpath("//div[@class='fm']"))->getText();

    $this->assertContains($subject,$getText);
    $this->assertContains($body,$getText);

    $this->driver->findElement(WebDriverBy::xpath("//div[contains(@class,'fm')]//div[@role='listitem'][1]"))->click();
    sleep(1);
    $this->driver->findElement(WebDriverBy::xpath("//div[contains(@class,'V j hj') and text()='Details']"))->click();
    $getSubject = $this->driver->findElement(WebDriverBy::xpath("//span[@class='kj']/span"))->getText();
    $getBody = $this->driver->findElement(WebDriverBy::xpath("//div[@class='Hi']"))->getText();
    $getSendEmail = $this->driver->findElement(WebDriverBy::xpath("//div[@class='Kg']/span"))->getText();
    $getToEmail = $this->driver->findElement(WebDriverBy::xpath("//div[@class='Kh']//span[@class='Hh']//span[@class='Kg ph']"))->getText();

    $this->assertContains($subject,$getSubject);
    $this->assertContains($body,$getBody);
    $this->assertContains($sendEmail,$getSendEmail);
    $this->assertContains($toEmail,$getToEmail);
  }

  protected function delete_sent_email()
  {
    $this->driver->get($this->SENT_EMAIL_URL);
    sleep(2);
    $this->driver->findElement(WebDriverBy::xpath("//div[@class='V j cb Ol'][1]"))->click();
    sleep(1);
    $this->driver->findElement(WebDriverBy::xpath("//div[@class='M j T b hc  Vm Im']//div[@class='V j Xd']"))->click();
    sleep(1);
    $noEmailSentEmail = $this->driver->findElement(WebDriverBy::xpath("//div[@class='Wl']"))->getText();

    $this->assertEquals($this->MSG_NOEMAIL,$noEmailSentEmail);
  }

  protected function delete_trash_email()
  {
    $this->driver->get($this->TRASH_URL);
    sleep(2);
    $this->driver->findElement(WebDriverBy::xpath("//div[@class='V j cb Ol'][1]"))->click();
    sleep(1);
    $this->driver->findElement(WebDriverBy::xpath("//div[@class='M j T b hc q m']//div[text()='Delete forever']"))->click();
    sleep(1);
    $noEmailTrash = $this->driver->findElement(WebDriverBy::xpath("//div[@class='Wl']"))->getText();

    $this->assertEquals($this->MSG_NOEMAIL,$noEmailTrash);
  }

  protected function delete_primary_email()
  {
    $this->driver->get($this->PRIMARY_EMAIL_URL);
    sleep(2);
    $this->driver->findElement(WebDriverBy::xpath("//div[@class='V j cb Ol'][1]"))->click();
    sleep(1);
    $this->driver->findElement(WebDriverBy::xpath("//div[@class='M j T b hc  Vm Im']//div[@class='V j Xd']"))->click();
    sleep(1);
    $noEmailPrimary = $this->driver->findElement(WebDriverBy::xpath("//div[@class='Wl']"))->getText();

    $this->assertContains($this->MSG_NOEMAIL_PRIMARY1,$noEmailPrimary);
    $this->assertContains($this->MSG_NOEMAIL_PRIMARY2,$noEmailPrimary);
  }

  public function tearDown() {
    if ($this->driver) {
      $this->driver->quit();
      sleep(10);
    }
    parent::tearDown();
  }
}
